core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java

```diff
@@ -0,0 +1,283 @@
+        put(Token[].class, TOKEN_ARRAYS);
+        put(ValueExpression.class, VALUE_EXPRESSIONS);
+        put(Expression.class, EXPRESSIONS);
+        put(Value.class, VALUES);
+        put(BinaryOperator.class, REDUCERS);
+        put(Slice.class, SLICES);
+        put(byte[].class, BYTE_ARRAYS);
+        put(Source.class, SOURCES);
+        put(long.class, LONGS);
+        put(int.class, INTEGERS);
+        put(ParseGraph.class, PARSEGRAPHS);
+        put(ParseItem.class, PARSEITEMS);
+        put(ByteStream.class, BYTESTREAMS);
+    }};
+
+    @Parameterized.Parameters(name="{0}")
+    public static Collection<Object[]> data() {
+        return generateObjectArrays(
+            Cho.class, Def.class, Nod.class, Opt.class, Pre.class, Rep.class, RepN.class, Seq.class, Sub.class,
+            Tie.class, TokenRef.class, While.class,
+            Len.class, Offset.class, Neg.class, Not.class, Count.class, First.class, Last.class, Reverse.class,
+            And.class, Or.class, ShiftLeft.class, ShiftRight.class, Add.class, Div.class, Mod.class, Mul.class,
+            io.parsingdata.metal.expression.value.arithmetic.Sub.class, Cat.class, Eq.class, EqNum.class, EqStr.class,
+            GtNum.class, LtNum.class, Nth.class, Elvis.class, io.parsingdata.metal.expression.logical.And.class,
+            io.parsingdata.metal.expression.logical.Or.class, FoldLeft.class, FoldRight.class, Value.class,
+            ParseValue.class, io.parsingdata.metal.expression.logical.Not.class, Const.class, NameRef.class,
+            io.parsingdata.metal.expression.value.reference.TokenRef.class, ConstantSource.class, Slice.class,
+            ParseReference.class, DataExpressionSource.class, ParseGraph.class, ByteStreamSource.class
+        );
+    }
+
+    private static Collection<Object[]> generateObjectArrays(Class... classes) {
+        Collection<Object[]> results = new ArrayList<>();
+        for (Class c : classes) {
+            results.add(generateObjectArrays(c));
+        }
+        return results;
+    }
+
+    private static Object[] generateObjectArrays(Class c) {
+        Constructor cons = c.getConstructors()[0];
+        List<List<Object>> args = new ArrayList<>();
+        for (Class cl : cons.getParameterTypes()) {
+            args.add(mapping.get(cl));
+        }
+        List<List<Object>> argLists = generateCombinations(0, args);
+        Object[] instances = new Object[3];
+        try {
+            instances[0] = cons.newInstance(argLists.get(0).toArray());
+            instances[1] = cons.newInstance(argLists.get(0).toArray());
+            List<Object> otherInstances = new ArrayList<>();
+            for (List<Object> argList : argLists.subList(1, argLists.size())) {
+                otherInstances.add(cons.newInstance(argList.toArray()));
+            }
+            instances[2] = otherInstances.toArray();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return instances;
```

Comment:
And as a side note, you could use the streaming API to replace some of the "old school" for loops in this code.

(But not the one right here, since `newInstance` throws an `Exception`.)